06-security.md

```diff
@@ -1 +1,451 @@
+
+### No technical details in error messages
+
+Stacktraces and vendor specific error codes tell a lot about the system and make informed guessing easier. If an error occurs these details should be logged but not presented to the user of the API. You can provide a timestamp as a reference for finding the stacktrace in the log. 
+
+### Do not repeat (erroneous) input in error messages
+
+In some conditions the browser may display the contents unescaped, potentially triggering cross site scripting. (Also this eliminates a lot of false positives when using vulnerability scanners).  
+
+### Restrict access to API documentation
+
+Unless the API is intended for public use, documentation should be available to developers only (Tools like [Swagger](http://swagger.io/) are useful when developing but should not be part of the production environment). 
+
+## Security on an API for Browser based consumption 
+
+This section will focus on the traditional session based authentication as most of our applications use a browser based user interface. 
+
+### Require Authentication on all API calls.
+
+The one notable exception being a GET request for the current authenticated user (if any). This can also be used to retrieve the XSRF token.
+
+### Don't use HTTP Basic Authentication
+
+When using Basic Authentication the credentials will be collected once by the browser and then sent along automatically on a header for each request.
+
+While simple and supported by almost everything this authentication method has some serious disadvantages:
+
+* the password is sent on each request.
+* the password is held in memory in the browser.
+* (header) logging may reveal the password.
+* there is no standard way of logging out once authenticated. [Clever Hacks](http://stackoverflow.com/questions/233507/how-to-log-out-user-from-web-site-using-basic-authentication) do exist but don't work on all browsers.
+
+### Don't use JSON Web Tokens (JWT) in a browser
+
+While JWT is perfect for machine to machine authentication, its very susceptible to theft using cross site scripting (XSS) in a browser. 
+Also JSON Web Tokens remain valid until they expire, so something equivalent to logging out is impossible without removing the statelessness advantage. 
+
+### Don't store tokens or personal information in session or local storage.
+
+These are easily read and tampered with by the user or a malicious script if XSS exists in the application.
+
+### GET requests must not change server state
+
+Of course this would be against the REST design principles elsewhere in this convention, but there is a good security reason too: they are easily executed using standard HTML tags.
```

Comment:
A good example is a Google crawler that only executes GET calls. You really don't want those to change your state :)